Mathlib/LinearAlgebra/FreeModule/Rank.lean

```diff
@@ -123,3 +137,109 @@ theorem rank_tensorProduct' (N : Type v) [AddCommGroup N] [Module R N] [Module.F
 #align rank_tensor_product' rank_tensorProduct'
 
 end CommRing
+
+section DivisionRing
+
+variable (K : Type u) [DivisionRing K]
+
+/-- Key lemma towards the Erdős-Kaplansky theorem from https://mathoverflow.net/a/168624 -/
+theorem rank_pi_nat : max ℵ₀ #K ≤ Module.rank K (ℕ → K) := by
+  have aleph0_le : ℵ₀ ≤ Module.rank K (ℕ → K) := (rank_finsupp_self K ℕ).symm.trans_le
+    (Finsupp.lcoeFun.rank_le_of_injective <| by exact FunLike.coe_injective)
+  refine max_le aleph0_le ?_
+  obtain card_K | card_K := le_or_lt #K ℵ₀
+  · exact card_K.trans aleph0_le
+  by_contra!
+  obtain ⟨⟨ιK, bK⟩⟩ := Module.Free.exists_basis (R := K) (M := ℕ → K)
+  let L := Subfield.closure (Set.range (fun i : ιK × ℕ ↦ bK i.1 i.2))
+  have hLK : #L < #K
+  · refine (Subfield.cardinal_mk_closure_le_max _).trans_lt
+      (max_lt_iff.mpr ⟨mk_range_le.trans_lt ?_, card_K⟩)
+    rwa [mk_prod, ← aleph0, lift_uzero, ← rank_eq_cardinal_basis' bK, mul_aleph0_eq aleph0_le]
+  letI := Module.compHom K (RingHom.op L.subtype)
+  obtain ⟨⟨ιL, bL⟩⟩ := Module.Free.exists_basis (R := Lᵐᵒᵖ) (M := K)
+  have card_ιL : ℵ₀ ≤ #ιL
+  · contrapose! hLK
+    haveI := @Fintype.ofFinite _ (lt_aleph0_iff_finite.mp hLK)
+    rw [bL.repr.toEquiv.cardinal_eq, mk_finsupp_of_fintype,
+        ← MulOpposite.opEquiv.cardinal_eq] at card_K ⊢
+    apply power_nat_le
+    contrapose! card_K
+    exact (power_lt_aleph0 card_K <| nat_lt_aleph0 _).le
+  obtain ⟨e⟩ := lift_mk_le'.mp (card_ιL.trans_eq (lift_uzero #ιL).symm)
+  have rep_e := bK.total_repr (bL ∘ e)
+  rw [Finsupp.total_apply, Finsupp.sum] at rep_e
+  set c := bK.repr (bL ∘ e)
+  set s := c.support
+  let f i (j : s) : L := ⟨bK j i, Subfield.subset_closure ⟨(j, i), rfl⟩⟩
+  have : ¬LinearIndependent Lᵐᵒᵖ f := fun h ↦ by
+    have := cardinal_lift_le_rank_of_linearIndependent' h
+    rw [lift_uzero, (LinearEquiv.piCongrRight fun _ ↦ MulOpposite.opLinearEquiv Lᵐᵒᵖ).rank_eq,
+        rank_fun'] at this
+    exact (nat_lt_aleph0 _).not_le this
+  obtain ⟨t, g, eq0, i, hi, hgi⟩ := not_linearIndependent_iff.mp this
+  refine hgi (linearIndependent_iff'.mp (bL.linearIndependent.comp e e.injective) t g ?_ i hi)
+  clear_value c s
+  simp_rw [← rep_e, Finset.sum_apply, Pi.smul_apply, Finset.smul_sum]
+  rw [Finset.sum_comm]
+  refine Finset.sum_eq_zero fun i hi ↦ ?_
+  replace eq0 := congr_arg L.subtype (congr_fun eq0 ⟨i, hi⟩)
+  rw [Finset.sum_apply, map_sum] at eq0
+  have : SMulCommClass Lᵐᵒᵖ K K := ⟨fun _ _ _ ↦ mul_assoc _ _ _⟩
```

Comment:
Shouldn't this SMulCommClass exist somewhere in mathlib? either as an instance, or at least
as something that can be invoked right away?

Reply:
The action of `Lᵐᵒᵖ` on `K` is itself obtained manually earlier in the proof rather than automatically:
```
letI := Module.compHom K (RingHom.op L.subtype)
```
So we can't expect Lean to know about the SMulCommClass here. I think mathlib currently lack actions by MulOpposite of subobjects and it feels like a big project to add them all, although actions by subobjects of MulOpposite should already be obtainable via typeclass inference, but we seem to be lacking a Subring analogue of [Subgroup.op](https://leanprover-community.github.io/mathlib4_docs/Mathlib/GroupTheory/Subgroup/MulOpposite.html#Subgroup.op).